Unix/http/httpclientauth.c

```diff
@@ -211,13 +217,224 @@ typedef struct _Gss_Extensions
 
     gss_OID Gss_Nt_Service_Name;
     gss_OID Gss_C_Nt_User_Name;
+
 } Gss_Extensions;
 
 static Gss_Extensions _g_gssClientState = { 0 };
 static struct _Once    g_once_state = ONCE_INITIALIZER;
 
 static const char GSS_LIBRARY_NAME[] = CONFIG_GSSLIB;
 
+
+
+/*
+ * Credentials are expected to live in the file ~/.omi/ntlmcred.
+ * If the environment variable NTLM_USER_FILE is set, it is respected.
+ * If not, the environment variable is set internally
+ * Either way, the directory the creds file is in must be user-only
+ *
+ * So: NTLM_USER_FILE is set? -> we use the file specified by it if the permissions are right.
+ *                   not set? -> we use ~/.omi/ntlmcred if the permissions are right.
+ *
+ * If neither condition is true, error
+ */
+
+static
+_Success_(return == 0) int _ValidateClientCredentials(HttpClient_SR_SocketData *context)
+{
+   char *ntlm_user_file = getenv("NTLM_USER_FILE");
+
+  static Probable_Cause_Data AUTH_ERROR_CRED_DIR = {
+              ERROR_ACCESS_DENIED,
+              WSMAN_CIMERROR_PROBABLE_CAUSE_AUTHENTICATION_FAILURE,
+              MI_T("credentials directory ~/.omi cannot be opened") 
+       };
+
+  static Probable_Cause_Data AUTH_ERROR_CRED_FILE = {
+              ERROR_ACCESS_DENIED,
+              WSMAN_CIMERROR_PROBABLE_CAUSE_AUTHENTICATION_FAILURE,
+              MI_T("credentials file and directory does not exist") 
+       };
+
+  static Probable_Cause_Data AUTH_ERROR_CRED_DIR_PERMS = {
+              ERROR_ACCESS_DENIED,
+              WSMAN_CIMERROR_PROBABLE_CAUSE_AUTHENTICATION_FAILURE,
+              MI_T("credentials file directory permissions must be user-only") 
+       };
+
+  static Probable_Cause_Data AUTH_ERROR_CRED_FILE_PERMS = {
+              ERROR_ACCESS_DENIED,
+              WSMAN_CIMERROR_PROBABLE_CAUSE_AUTHENTICATION_FAILURE,
+              MI_T("credentials file permissions must be user-only") 
+       };
+
+
+    if (ntlm_user_file)
+    {
+        // Check the permissions on the directory. They need to be read-only owner
+        
+        {
+            struct stat buf = {0};
+            char *ntlm_cred_dir = strdup(ntlm_user_file);
+            char* p = strrchr(ntlm_cred_dir, '/');
+            if (p)
+            {
+                *p = '\0';
+            }
+
+            int rtn = stat(ntlm_cred_dir, &buf);
+            if (rtn < 0)
+            {
+                free(ntlm_cred_dir);
+                _g_gssClientState.probableCause = &AUTH_ERROR_CRED_FILE;
+                goto Err1;
+               
+            }
+
+            if (!S_ISDIR(buf.st_mode))
+            {
+               // Not a directory? Unlikely, but complain and issue error
+                free(ntlm_cred_dir);
```

Comment:
you already have an error path, why not free once in there rather than duplicate in every error case?